File: examples/python/auth_and_errors.py
"""Copyright 2026 Volstrata.com

Authentication, refusals and rate limits -- the three things a client has to get
right before anything else matters.

Docs: https://volstrata.com/docs/api-auth
      https://volstrata.com/docs/api-errors

    python auth_and_errors.py

Runs identically with or without a key, and exits 0 either way. Five things
happen, in order:

  1. an anonymous call to a capability that answers without a credential
  2. an authenticated call, when VOLSTRATA_API_KEY is set        (or skipped)
  3. a call to a capability above your floor       (refused, caught, explained)
  4. a call to a path that does not exist                 (404, caught)
  5. the rate-limit headers, read off a real response

A refusal is data, not a crash. That is the whole lesson of this file.
"""

import sys

from volstrata_helpers import (
    VolstrataError,
    call,
    get,
    have_key,
    heading,
    rate_limit,
    row,
    run_examples,
    show,
)


def anonymous_call() -> None:
    """Much of the published surface answers with no credential at all.

    Nothing is sent here but a User-Agent -- which is not optional. Some default
    agent strings are refused at the CDN before the API ever sees the request,
    and that refusal is plain text, not problem+json. If you get a non-JSON
    error body, check your User-Agent first.
    """
    heading("1. Anonymous call -- no credential sent")
    data = get("/gex/levels", ticker="SPX")
    row("ok", data.get("ok"))
    row("ticker", data.get("ticker"))
    row("spot", data.get("spot"))
    print("  That is the whole setup: one GET, no account.")


def authenticated_call() -> None:
    """A key is presented as `Authorization: Bearer <key>`.

    The alternative is `?api_key=<key>` on the query string, accepted everywhere
    the header is. Prefer the header: query strings leak into access logs,
    proxies and browser history.

    A key is issued once and shown in full once, at creation; the service keeps
    only a hash. Personal access tokens and service-account tokens are accepted
    by the same header -- the 401 body in step 3 names the prefixes it takes.

    Issue and revoke keys at https://volstrata.com/api-keys
    Guidance: https://volstrata.com/docs/api-keys
    """
    heading("2. Authenticated call")
    if not have_key():
        print("  Skipped: VOLSTRATA_API_KEY is not set.")
        print("  Export it and re-run to see this branch:")
        print("    export VOLSTRATA_API_KEY=...        # macOS / Linux")
        print("    $env:VOLSTRATA_API_KEY = \"...\"      # Windows PowerShell")
        return

    # meta.access reports what the *calling credential* is entitled to, which
    # makes it the fastest way to confirm a key is being read correctly.
    access = get("/meta/access")
    row("ok", access.get("ok"))
    for field in ("caller_tier", "unlocked_count", "locked_count", "total",
                  "key_rate_per_min", "limit_cap"):
        if field in access:
            row(field, access[field])
    print("  caller_tier is 'anon' when no credential arrived. If you set a key")
    print("  and still see 'anon', the header is not reaching the API -- check")
    print("  that before anything else.")


def gated_call() -> None:
    """A capability above your floor is refused with a body that says so.

    Two distinct refusals live here, and telling them apart matters:

      401 `auth_required` -- no usable credential was presented at all. The
          `detail` names the credential families the API accepts.
      402 with `required_plan_name` -- the credential is valid, the plan is not
          high enough. The body adds `feature`, `required_plan`,
          `required_plan_name` and `current_plan`.

    gex.snapshot sits above the lowest floor, so anonymously you get the first
    and with a key below its floor you get the second. This prints whichever one
    you actually received.
    """
    heading("3. A capability above the anonymous floor -- gex.snapshot")
    try:
        data = get("/gex/snapshot", ticker="SPX")
    except VolstrataError as exc:
        row("status", exc.status)
        row("code", exc.code)
        row("title", exc.title)
        row("detail", exc.detail)
        if exc.needs_plan:
            row("required_plan_name", exc.required_plan_name)
            row("current_plan", exc.current_plan)
            print("  -> Nothing about the request was wrong. The account is below")
            print("     the plan floor, and the body names the plan that clears it.")
        else:
            print("  -> No usable credential was presented. Set VOLSTRATA_API_KEY and")
            print("     re-run; with a key below the floor this becomes a 402 that")
            print("     names the plan you would need.")
        print("  Handle this branch in your own client, and do not retry it: the")
        print("  answer will not change and the attempt still costs rate budget.")
        return
    row("ok", data.get("ok"))
    print("  Your plan includes this one, so it returned normally.")


def bad_path() -> None:
    """A path that does not exist answers with the same problem+json envelope.

    `route_not_found` means the URL is wrong. Plain `not_found` means the route
    was right and the resource is not available just now -- a distinction worth
    honouring in your error handling, because only one of them is a bug in your
    code. Every published path is listed by GET /api/v1/meta/capabilities.
    """
    heading("4. A path that does not exist")
    try:
        get("/gex/not_a_real_capability")
    except VolstrataError as exc:
        row("status", exc.status)
        row("code", exc.code)
        row("title", exc.title)
        row("detail", exc.detail)
        row("request_id", "present" if exc.request_id else "absent")
        print("  Every refusal carries a request_id. Quote it in a support request")
        print("  and the call can be traced.")
        return
    print("  Unexpected: that path answered successfully.")


def rate_limit_headers() -> None:
    """Rate limits are per owner, not per key.

    Minting a second key does not raise your ceiling; the budget belongs to the
    account and scales with the plan. Two header families are sent on every
    response -- the modern combined `ratelimit` header plus `ratelimit-policy`,
    and the older `x-ratelimit-*` set. `ratelimit-policy` lists each policy in
    play and marks which one is enforced, and `x-ratelimit-bucket` names the
    bucket you are counted in.

    On a 429 the body is `rate_limited` and `Retry-After` says how many seconds
    to wait. volstrata_helpers.request_with_backoff() honours it, retries only
    429 and 5xx, and gives up after four attempts.

    Reference: https://volstrata.com/docs/api-rate-limits
    """
    heading("5. Rate-limit headers")
    response = call("/gex/maxpain", params={"ticker": "SPX"})
    show(rate_limit(response))
    print("  Read `remaining` and slow down before you are told to. Everything in")
    print("  this repo is sequential for the same reason.")


def main() -> int:
    print("VolStrata: auth, errors and rate limits -- https://volstrata.com")
    exit_code = run_examples((
        ("anonymous call", anonymous_call),
        ("authenticated call", authenticated_call),
        ("gated call", gated_call),
        ("bad path", bad_path),
        ("rate-limit headers", rate_limit_headers),
    ))

    heading("Done")
    print("  Every branch above finished cleanly, including the two failures.")
    print("  Error reference: https://volstrata.com/docs/api-errors")
    return exit_code


if __name__ == "__main__":
    sys.exit(main())

File: examples/javascript/auth_and_errors.mjs
/**
 * Copyright 2026 Volstrata.com
 *
 * Authentication and the error envelope, in one runnable file.
 * Docs: https://volstrata.com/docs/api-errors
 *
 * Run:  node auth_and_errors.mjs
 *       npm run auth
 *
 * Five things happen, in order:
 *
 *   1. a Free call with no credential at all      → 200
 *   2. the same call with a key, when one is set  → 200 (skipped otherwise)
 *   3. a capability above the Free floor          → 402, naming the plan needed
 *   4. a path that does not exist                 → 404 route_not_found
 *   5. the rate-limit headers every response carries
 *
 * Every branch is handled, so this file exits 0 whether or not you have a key.
 * Three requests are made, or four with a key set.
 */

import {
  BASE,
  TIMEOUT_MS,
  USER_AGENT,
  VolstrataApiError,
  apiFetch,
  buildUrl,
  hasApiKey,
  heading,
  keysOf,
  show,
  showRateLimit,
} from "./lib/volstrata.mjs";

// ---------------------------------------------------------------------------
// 1. No credential at all
// ---------------------------------------------------------------------------
//
// 67 of the 180 published operations sit at the Free plan floor and most of
// those, `gex.levels` included, answer a caller who sends nothing at all. This
// is a supported mode, not a degraded one, so the request below does NOT go
// through the shared helper — it sends only the two headers every HTTP client
// should send, and no Authorization header even if you have a key exported.

heading("1. Anonymous — gex.levels (Free)");

let anonymousResponse = null;
try {
  const url = buildUrl("/api/v1/gex/levels", { ticker: "SPX" });
  const res = await fetch(url, {
    headers: {
      // Mandatory, not politeness. A CDN sits in front of the API and refuses
      // some default agents outright; that refusal is HTML or plain text and
      // never reaches the API, so it cannot be an API error however it looks.
      "User-Agent": USER_AGENT,
      Accept: "application/json",
    },
    signal: AbortSignal.timeout(TIMEOUT_MS),
  });

  // Status first. Only then the body.
  if (!res.ok) throw await VolstrataApiError.fromResponse(res, url);
  const data = await res.json();

  anonymousResponse = res;
  show("status", res.status);
  show("ticker", data.ticker);
  show("spot", data.spot);
  show("levels", keysOf(data.levels, 8));
} catch (error) {
  reportError(error);
}

// ---------------------------------------------------------------------------
// 2. With a key
// ---------------------------------------------------------------------------
//
// One header carries it:
//
//     Authorization: Bearer <your key>
//
// The API also accepts `?api_key=<your key>` on the query string, which is
// occasionally the only option (a spreadsheet, a chart tool). Prefer the header:
// query strings end up in shell history, proxy logs and referrer headers.
//
// Keys look like `gex_key_v1_` followed by 43 url-safe base64 characters. The
// older `sig_key_v1_` prefix is still accepted. Two other credential families —
// personal access tokens (`sig_pat_v1_`) and service accounts (`sig_svc_v1_`) —
// are accepted on the same header; this repo teaches API keys.
//
// Never put any of them in source. `lib/volstrata.mjs` reads VOLSTRATA_API_KEY
// from the environment and attaches the header only when it is present.

heading("2. Authenticated — same capability, with a key");

if (!hasApiKey) {
  console.log("  VOLSTRATA_API_KEY is not set, so this step is skipped.");
  console.log("  Nothing above needed it. Get one at https://volstrata.com/api-keys,");
  console.log("  then:");
  console.log('    export VOLSTRATA_API_KEY="gex_key_v1_XXXXXXXX"      # macOS / Linux');
  console.log('    $env:VOLSTRATA_API_KEY = "gex_key_v1_XXXXXXXX"      # Windows PowerShell');
} else {
  try {
    const { res, data } = await apiFetch("/api/v1/gex/levels", { ticker: "SPX" });
    show("status", res.status);
    show("ticker", data.ticker);
    show("auth", "sent as Authorization: Bearer …");
    // A key raises the ceiling and unlocks capabilities; the response shape of
    // a Free capability is the same either way.
    showRateLimit(res);
  } catch (error) {
    reportError(error);
  }
}

// ---------------------------------------------------------------------------
// 3. A capability above your plan floor
// ---------------------------------------------------------------------------
//
// `gex.snapshot` requires the Edge plan, so this step refuses — in one of two
// different ways, and telling them apart is the point of this section.
//
// With no credential you get 401 `auth_required`, and the detail names the
// credential families the API accepts:
//
//     {"ok": false, "status": 401, "code": "auth_required",
//      "type": "<origin>/errors/auth_required",
//      "title": "Authentication required", "detail": "…",
//      "instance": "/api/v1/gex/snapshot", "request_id": "…"}
//
// With a valid credential whose plan sits below the floor, the credential is
// fine and the plan is not, so the problem document carries four extra fields:
//
//     "feature": "gex.snapshot", "required_plan": "edge",
//     "required_plan_name": "Edge", "current_plan": "…"
//
// Show `required_plan_name` to a human ("Edge") and branch on `required_plan`
// or `code` in software. Never retry either refusal: it will be refused again
// just as fast, and the attempt still counts against your rate limit.

heading("3. Gated — gex.snapshot (Edge)");

try {
  const { data } = await apiFetch("/api/v1/gex/snapshot", { ticker: "SPX" });
  show("status", 200);
  show("keys", keysOf(data, 12));
  show("note", "your plan includes this capability");
} catch (error) {
  if (error instanceof VolstrataApiError && error.isPlanRefusal) {
    // Authenticated, but below the floor.
    show("status", error.status);
    show("refused", `needs the ${error.requiredPlanName} plan`);
    show("code", error.code);
    show("current plan", error.currentPlan ?? "(unknown)");
    show("request_id", error.requestId ?? "(none)");
    console.log("\n  Plans: https://volstrata.com/docs/api-auth");
  } else if (error instanceof VolstrataApiError && error.isAuthRequired) {
    // No credential at all — the usual outcome when running this file with no
    // key. A key gets you past this; whether the data then comes back depends
    // on the plan behind it.
    show("status", error.status);
    show("code", error.code);
    show("detail", error.detail ?? "(none)");
    show("request_id", error.requestId ?? "(none)");
    console.log("\n  Set VOLSTRATA_API_KEY and run again to see the plan-floor form");
    console.log("  of this refusal: https://volstrata.com/docs/api-auth");
  } else {
    reportError(error);
  }
}

// ---------------------------------------------------------------------------
// 4. A path that does not exist
// ---------------------------------------------------------------------------
//
// Worth seeing once so you can tell it apart from a plan refusal: a wrong path
// is 404 `route_not_found`, and no key or plan will ever fix it. Capability
// names are dotted (`gex.levels`) and the path is the same name with slashes
// (`/api/v1/gex/levels`) — the two are always in step.

heading("4. Wrong path — 404 route_not_found");

try {
  await apiFetch("/api/v1/gex/no-such-capability");
  show("status", "200 — unexpected; that path is not published");
} catch (error) {
  if (error instanceof VolstrataApiError && error.isApiResponse) {
    show("status", error.status);
    show("code", error.code);
    show("title", error.title);
    show("detail", error.detail ?? "(none)");
    show("instance", error.instance ?? "(none)");
  } else {
    reportError(error);
  }
}

// ---------------------------------------------------------------------------
// 5. Rate limits
// ---------------------------------------------------------------------------
//
// Every response carries both header families: the modern combined
// `ratelimit: limit=…, remaining=…, reset=…` with `ratelimit-policy`, and the
// legacy `x-ratelimit-limit` / `-remaining` / `-reset`. Read whichever suits
// your stack — they describe the same budget.
//
// The budget is per OWNER, not per key: minting a second key does not raise the
// ceiling. Anonymous callers get the smallest one and it rises with the plan —
// read `remaining` and `reset` off the response rather than hard-coding a
// number. On a 429, `Retry-After` says how long to wait; honour it instead of
// inventing a delay. `withBackoff()` in lib/volstrata.mjs does exactly that.

heading("5. Rate-limit headers");

if (anonymousResponse) {
  showRateLimit(anonymousResponse);
} else {
  console.log("  (no response captured above)");
}
console.log("\n  Reference: https://volstrata.com/docs/api-rate-limits");

heading("Done");
console.log(`  base url   ${BASE}`);
console.log(`  api key    ${hasApiKey ? "detected" : "not set — every step above still ran"}`);
console.log("  exit code  0 in every branch: a refusal is an answer, not a crash.");

/** Print any failure without ever failing the run. */
function reportError(error) {
  if (error instanceof VolstrataApiError) {
    console.log(error.describe("  "));
    if (!error.isApiResponse) {
      console.log("  This response did not come from the API — see the note in step 1.");
    }
  } else {
    console.log(`  request failed: ${error?.message ?? error}`);
  }
}
